curvefs/src/metaserver/s3compact_manager.cpp

```diff
@@ -0,0 +1,107 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * @Project: curve
+ * @Date: 2021-09-07
+ * @Author: majie1
+ */
+
+#include "curvefs/src/metaserver/s3compact_manager.h"
+
+#include <thread>
+
+#include "curvefs/src/metaserver/s3compact_wq_impl.h"
+
+using curve::common::Configuration;
+using curve::common::InitS3AdaptorOption;
+using curve::common::ReadLockGuard;
+using curve::common::S3Adapter;
+using curve::common::S3AdapterOption;
+using curve::common::TaskThreadPool;
+using curve::common::WriteLockGuard;
+
+namespace curvefs {
+namespace metaserver {
+
+S3CompactManager::Init(std::shared_ptr<Configuration> conf,
+                       std::shared_ptr<CopysetNodeManager> copysetNodeManager) {
+    conf_ = conf;
+    copysetNodeManager_ = copysetNodeManager;
+    // start a s3 client
+    s3Adapter_ = std::make_shared<S3Adapter>();
+    S3AdapterOption option;
+    InitS3AdaptorOption(conf_.get(), &option);
+    s3Adapter_->Init(option);
+    // init workqueue
+    s3compactworkqueueImpl_ =
+        std::make_shared<S3CompactWorkQueueImpl>(s3Adapter_, conf_);
+    inited = true;
+}
+
+void S3CompactManager::RegisterS3Compact(std::shared_ptr<S3Compact> s3compact) {
+    WriteLockGuard l(rwLock_);
+    s3compacts_.emplace_back(s3compact);
+}
+
+void S3CompactManager::Enqueue() {
+    std::vector<S3Compact> s3compacts;
+    {
+        // make a snapshot
+        ReadLockGuard l(rwLock_);
+        s3compacts = s3compacts_;
+    }
+    for (const auto& s3compact : s3compacts) {
+        // question: need to check pointer validation?
+        auto pinfo = s3compact->GetPartition();
+        auto copysetNode = copysetNodeManager_->GetCopysetNode(
+            pinfo.poolId(), pinfo.copysetId());
+        // traverse inode container
+        auto inodeStorage = s3compat->GetMutableInodeStorage();
+        InodeContainerType inodes(*(inodeStorage->GetInodeContainerType()));
+        for (const auto& inode : inodes) {
+            // am i copyset node leader?
+            if (!copysetNode->IsLeaderTerm()) continue;
+            s3compactworkqueueImpl_->Enqueue(inodeStorage, InodeKey(inode),
+                                             pinfo, copysetNode);
+        }
+    }
+}
+
+int S3CompactManager::Run() {
+    if (!inited) return -1;
+    int threadNum;
+    int queueSize;
+    LOG_IF(FATAL, !conf_->GetIntValue("s3compactwq.thread_num", &threadNum));
+    LOG_IF(FATAL, !conf_->GetIntValue("s3compactwq.queue_size", &queueSize));
+    int ret = s3compactworkqueueImpl_->Start(threadNum, queueSize);
+    if (ret != 0) return ret;
+    std::thread entry([this]() {
```

Comment:
it's better to let `entry` as a member variable of `S3CompactManager`, and Stop it when Manager stops

Reply:
fixed.